roles/maas_nameserver/tasks/main.yml

```diff
@@ -0,0 +1,117 @@
+---
+# Include secrets from maas.yml (e.g., maas_api_key, maas_api_url).
+# Note: dns_domains should be defined in the inventory's group_vars/all.yml
+- name: Include secrets
+  ansible.builtin.include_vars: "{{ item }}"
+  no_log: true
+  with_first_found:
+    - "{{ secrets_path | mandatory }}/maas.yml"
+  tags:
+    - always
+
+- name: Ensure MAAS CLI is logged in
+  ansible.builtin.command: "maas login {{ maas_profile }} {{ maas_api_url }} {{ maas_api_key }}"
+  register: maas_login
+  changed_when: maas_login.rc == 0
+  failed_when: maas_login.rc != 0 and maas_login.stderr != "Profile already exists"
+
+- name: Get existing DNS resources
+  ansible.builtin.command: "maas {{ maas_profile }} dnsresources read"
+  register: existing_resources
+  changed_when: false
+
+- name: Initialize DNS records list
+  ansible.builtin.set_fact:
+    dns_records: []
+
+# Define target hosts list for DNS records, excluding the 'maas' group.
+- name: Initialize target hosts list
+  ansible.builtin.set_fact:
+    target_hosts_list: []
+
+- name: Build target hosts list
+  ansible.builtin.set_fact:
+    target_hosts_list: "{{ target_hosts_list + item.value }}"
+  loop: "{{ groups | dict2items }}"
+  when: item.key not in exclude_groups
+
+# After population, target_hosts will contain all hosts from inventory groups except 'maas', e.g.,
+# ['machine001.internal.example.com', 'snipe.internal.example.com', 'vm-14.example.com'].
+# This can be overridden in secrets/maas.yml (e.g., target_hosts: ['host1.example.com', 'host2.example.com'])
+# or via extra vars (e.g., -e "target_hosts=['host1.example.com']").
+- name: Define target hosts for DNS records
+  ansible.builtin.set_fact:
+    target_hosts: "{{ target_hosts if target_hosts | length > 0 else (target_hosts_list | flatten | unique | default([])) }}"
+  when: groups.keys() | length > 0
+
+- name: Build DNS records for all interfaces
+  ansible.builtin.set_fact:
+    dns_records: "{{ dns_records + [{'name': item[0].split('.')[0], 'ip': interface_ip, 'type': 'A', 'domain': item[1].value}] }}"
+  loop: "{{ (target_hosts | default([])) | product(dns_domains | dict2items) | list }}"
```

Comment:
I don't have the faintest idea what's going on here.  It looks as though this will make a list of two-element combinations of every item in target_hosts and every item in dns_domains.  How is that remotely useful?

I think at this point I'm going to have to stop reviewing until we can talk.  I'll save this review this far, but I'm not done.